### PR TITLE
dbcache phase 3 enhancements

### PR DESCRIPTION
**Pull request by ptschip**

### CODE REVIEW

File: doc/release-notes.md [hunk 1]
```diff
@@ -115,6 +115,7 @@ The p2p alert system has been removed in #7692 and the 'alert' message is no lon
 
 ### Miscellaneous
 
+
```
Comment (sickpig) [inline doc/release-notes.md]: unrelated change

**Comment by sickpig** [comment]
guess that a list of Core PR ported with this PR would be handy, so far this what I was able to recognize. 

https://github.com/bitcoin/bitcoin/pull/9902   ("Lightweight abstraction of boost::filesystem")
https://github.com/bitcoin/bitcoin/pull/10195 ("Switch chainstate db and cache to per-txout model")
https://github.com/bitcoin/bitcoin/pull/10445 ("Fix and add test for empty chain and reorg consistency for gettxoutsetinfo")

I'll add other in case I find more. 

edit: 

https://github.com/bitcoin/bitcoin/pull/8020 ("Use SipHash-2-4 for various non-cryptographic hashes")
https://github.com/bitcoin/bitcoin/pull/7927 ("Minor changes to dbwrapper to simplify support for other databases")
https://github.com/bitcoin/bitcoin/pull/9039: ("Various serialization simplifcations and optimizations")
https://github.com/bitcoin/bitcoin/pull/8914 ("Kill insecure_random and associated global state")
https://github.com/bitcoin/bitcoin/pull/11011 ("[Trivial] Add a comment on the use of prevector in script")

guess we need to add also: 

https://github.com/bitcoin/bitcoin/pull/11029: ("[RPC] trivial: gettxout no longer shows version of tx") relate to 10195 
https://github.com/bitcoin/bitcoin/pull/9792: ("FastRandomContext improvements and switch to ChaCha20") related to 8194

**Comment by sickpig** [comment]
@ptschip this a massive amount of changes wonder if we could find a way to split the PR somehow just to make the review easier, what do you think?

File: src/streams.h [hunk 2]
```diff
@@ -23,7 +24,52 @@
 #include <utility>
 #include <vector>
 
-#include "util.h"
+template<typename Stream>
+class OverrideStream
```
Comment (sickpig) [inline src/streams.h]: changes from L28 to L72 seems to come from this commit: 

```
commit 7030d9eb47254499bba14f1c00abc6bf493efd91
Author: Pieter Wuille <pieter.wuille@gmail.com>
Date:   Fri Nov 6 01:32:04 2015 +0100

    BIP144: Serialization, hashes, relay (sender side)

    Contains refactorings by Eric Lombrozo.
    Contains fixup by Nicolas Dorier.
    Contains cleanup of CInv::GetCommand by Alex Morcos
```

and seem not be used anywhere in the code, a left over? 

(such commit is included in a larer batch that come in via #8149)

**Comment by ptschip** [comment]
i should put together a list of the PR#'s i pulled in

**Comment by ptschip** [comment]
basically everything up to  392f89e (fix some empty vector references) 
is a pre-req.    (also, 0ec4173 to 21c67e1 as also pre-reqs that i missed)

I don't think breaking up will make any difference...the first half of
the commits are basically pre-reqs and the second half the meat of the
UTXO upgrade.

**Comment by ptschip** [comment]
@sickpig here's a list of what is in this PR and all the cherry's 

Pre-requesite for serialization.h changes.
PR https://github.com/bitcoin/bitcoin/pull/9039

Only one commit from this PR: Introduce FastRandomContext:: randbool()
  https://github.com/bitcoin/bitcoin/pull/9792

Prevent integer overflow in ReadVarint
https://github.com/bitcoin/bitcoin/pull/9693
  - not a pre-req but came across it as something we needed to pull in

https://github.com/bitcoin/bitcoin/pull/7933 - the following three commits only
  introduce constant for maximum CScript length
  Treat overly long scriptPubKeys as unspendable
  Fix OOM bug: UTXO entries with invalid script length
     - not a pre-req but came across it as something we needed to pull in

https://github.com/bitcoin/bitcoin/pull/10313
   Add constant for maximum stack size
  - not a pre-req but came across it as something we needed to pull in

https://github.com/bitcoin/bitcoin/pull/10534
  clarify prevector::erase and avoid swap to clear

https://github.com/bitcoin/bitcoin/pull/9902
   lightweight abstraction of boost::filesystem

https://github.com/bitcoin/bitcoin/pull/9895
   turn trycreate directory in to trycreatedirectories

https://github.com/bitcoin/bitcoin/pull/8020
   Use SipHash 2-4 for various non-cryptographic hashes

https://github.com/bitcoin/bitcoin/pull/7927
   Minor changes to dbwrapper to simplify support of other databases

https://github.com/bitcoin/bitcoin/pull/9346
  Batch construct batches

https://github.com/bitcoin/bitcoin/pull/7756
  Add cursor to iterate over UTXO set

https://github.com/bitcoin/bitcoin/pull/10250
  Fix some empty vector references


=========================================================

This PR is the meat of the utxo upgrade and has about 30 commits in it...
https://github.com/bitcoin/bitcoin/pull/10195
=========================================================

Then we add "Force on the fly compaction" which helps get rid of the old UTXO as we're
building the new, otherwise, we end up with a UTXO with double the size
https://github.com/bitcoin/bitcoin/pull/10526

Then finally just to fix a test
https://github.com/bitcoin/bitcoin/pull/10445


The only thing I added of any significance other than fixing tests was this commit
https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/785/commits/3b08aec8b25b63b8b77d38e9a7cb9d77f72ac137
  - Create random uint64_t by adding two 32 bit integers

**Comment by ptschip** [comment]
@sickpig you're right, that's a bit of dead code there..i'll remove and push up in a minute.

File: src/key.h [hunk 3]
```diff
@@ -170,6 +170,23 @@ struct CExtKey {
     bool Derive(CExtKey& out, unsigned int nChild) const;
     CExtPubKey Neuter() const;
     void SetMaster(const unsigned char* seed, unsigned int nSeedLen);
+    template <typename Stream>
```
Comment (gandrewstone) [inline src/key.h:173]: In the Encode and Decode functions above, we should use the BIP32_EXTKEY_SIZE constant, since the code below requires the proper array size.  Also we should change that constant to an enum since its defined in a header file.

File: src/primitives/block.h [hunk 4]
```diff
@@ -27,7 +27,8 @@ class CBlockHeader
 {
 public:
     // header
-    static const int32_t CURRENT_VERSION=BASE_VERSION;
+   // static const int32_t CURRENT_VERSION = BASE_VERSION;
```
Comment (gandrewstone) [inline src/primitives/block.h]: delete repeated line that's commented out

**Comment by gandrewstone** [comment]
thanks for this great porting effort!
